File: src/java/magma/collect/stream/HeadedStream.java
package magma.collect.stream;

import magma.collect.Collector;
import magma.option.Tuple;
import magma.result.Ok;
import magma.result.Result;

import java.util.function.BiFunction;
import java.util.function.Function;

public record HeadedStream<T>(Head<T> head) implements Stream<T> {
    @Override
    public <R> R fold(R initial, BiFunction<R, T, R> folder) {
        R current = initial;
        while (true) {
            R finalCurrent = current;
            Tuple<Boolean, R> tuple = head.next()
                    .map(inner -> folder.apply(finalCurrent, inner))
                    .toTuple(current);

            if (tuple.left()) {
                current = tuple.right();
            } else {
                break;
            }
        }

        return current;
    }

    @Override
    public <R> Stream<R> map(Function<T, R> mapper) {
        return new HeadedStream<>(() -> head.next().map(mapper));
    }

    @Override
    public <C> C collect(Collector<T, C> collector) {
        return fold(collector.createInitial(), collector::fold);
    }

    @Override
    public <R, X> Result<R, X> foldToResult(R initial, BiFunction<R, T, Result<R, X>> folder) {
        return this.<Result<R, X>>fold(new Ok<>(initial),
                (rxResult, t) -> rxResult.flatMapValue(inner -> folder.apply(inner, t)));
    }
}

File: src/clang/magma/collect/stream/head/HeadedStream.c
#include "HeadedStream.h"
magma.collect.stream.head.R foldWithInitial(magma.collect.stream.head.R initial, magma.collect.stream.head.R(*folder)(magma.collect.stream.head.R, T)){R current = initial;while (true) {
            R finalCurrent = current;
            Tuple<Boolean, R> tuple = head.next()
                    .map(inner -> folder.apply(finalCurrent, inner))
                    .toTuple(current);

            if (tuple.left()) {
                current = tuple.right();
            } else {
                break;
            }
        }return current;
}
magma.collect.stream.Stream<magma.collect.stream.head.R> map(magma.collect.stream.head.R(*mapper)(T)){return (__lambda0__.next().map(mapper));
}
magma.collect.stream.head.C collect(magma.collect.stream.Collector<T, magma.collect.stream.head.C> collector){return foldWithInitial(collector.createInitial(), collector.fold);
}
magma.option.Option<T> next(){return head.next();
}
int anyMatch(int(*predicate)(T)){return foldWithInitial(false, __lambda1__||predicate.test(t));
}
magma.option.Option<magma.collect.stream.head.R> foldMapping(magma.collect.stream.head.R(*mapper)(T), magma.collect.stream.head.R(*folder)(magma.collect.stream.head.R, T)){return head.next().map(mapper).map(__lambda2__(initial, folder));
}
magma.collect.stream.Stream<T> filter(int(*predicate)(T)){return flatMap(__lambda3__);
}
magma.collect.stream.Stream<magma.collect.stream.head.R> flatMap(magma.collect.stream.Stream<magma.collect.stream.head.R>(*mapper)(T)){return this.foldWithInitial((()), __lambda4__.concat(mapper.apply(t)));
}
magma.result.Result<magma.collect.stream.head.R, magma.collect.stream.head.X> foldToResult(magma.collect.stream.head.R initial, magma.result.Result<magma.collect.stream.head.R, magma.collect.stream.head.X>(*folder)(magma.collect.stream.head.R, T)){return this.foldWithInitial((initial), __lambda5__.flatMapValue(__lambda6__.apply(inner, t)));
}
magma.collect.stream.Stream<T> concat(magma.collect.stream.Stream<T> other){return (__lambda7__.next().or(other.next));
}
auto __lambda0__();
auto __lambda1__();
auto __lambda2__();
auto __lambda3__();
auto __lambda4__();
auto __lambda5__();
auto __lambda6__();
auto __lambda7__();

